examples/graphrnn_s.py

```diff
@@ -0,0 +1,364 @@
+
+import torch_geometric
+from torch_geometric import data, loader
+from torch_geometric import transforms as T
+
+
+class EncodeGraphRNNFeature(T.BaseTransform):
+    r"""Encode the graph adjacency matrix into a sequence of adjacency vectors
+    of length :obj:`M`.
+    This is the approach described in the "GraphRNN: Generating Realistic
+    Graphs with Deep Auto-regressive Models"
+    <https://arxiv.org/pdf/1802.08773>`_ paper.
+
+    Args:
+        M (int): The length of the adjacency vectors.
+    """
+    def __init__(self, M):
+        self.M = M
+
+    @staticmethod
+    def extract_bands(adj, M):
+        r"""Uses stride tricks to extract :obj:`M`-long bands above the
+        diagonal of the given square matrix.
+
+        Args:
+            adj (Tensor): The adjacency matrix of shape :obj:`(N, N)`.
+            M (int): The number of bands to extract.
+
+        :rtype: Tensor
+        """
+        N = adj.shape[1]
+        padded_adj = torch.zeros((N + M - 1, N))
+        padded_adj[M - 1:, :] = adj
+        return padded_adj.as_strided(size=(N - 1, M), stride=(N + 1, N),
+                                     storage_offset=1)
+
+    @staticmethod
+    def bands_to_matrix(bands):
+        r"""Given M bands above the diagonal of a square matrix, return the
+        full matrix.
+
+        Args:
+            bands (Tensor): The M bands of shape :obj:`(N, M)`.
+
+        :rtype: Tensor
+        """
+        M = bands.shape[1]
+        N = bands.shape[0] + 1
+        padded_adj = torch.zeros((N + M - 1, N))
+        view = padded_adj.as_strided(size=(N - 1, M), stride=(N + 1, N),
+                                     storage_offset=1)
+        view[:, :] = bands
+        return padded_adj[M - 1:, :]
+
+    @staticmethod
+    def inverse(y):
+        r"""Inverse of the :obj:`__call__` method: given the encoded sequence
+        y, return the adjacency matrix.
+
+        Args:
+            y (Tensor): The encoded sequence of shape :obj:`(N, M)`.
+
+        :rtype: Tensor
+        """
+        bands = torch.flip(y, dims=[1])
+        adj = EncodeGraphRNNFeature.bands_to_matrix(bands)
+        return adj
+
+    def __call__(self, data):
+        adj = torch_geometric.utils.to_dense_adj(data.edge_index)
+        sequences = torch.flip(self.extract_bands(adj, self.M), dims=[1])
+
+        # Add SOS (row of ones) and EOS (row of zeros).
+        sequences = torch.cat(
+            [torch.ones(1, self.M), sequences,
+             torch.zeros(1, self.M)], dim=0)
+
+        data.length = data.num_nodes
+
+        data.x = sequences[:-1]
+        data.y = sequences[1:]
+        return data
+
+
+class BFS(T.BaseTransform):
+    r"""Start a breath first search from a random node and reorder the edge
+    list so that the node indices correspond to the breadth-first search order.
+    """
+    def __call__(self, data):
+        x = data.x
+        edge_index = data.edge_index
+        assert (data.is_undirected()
```

Comment:
just FYI I noticed that `EncodeGraphRNNFeature` works with an undirected graph right now, but I don't know if it makes sense in this case. Maybe its worth having the same assert there if not.